File: app/library.pl
use strict;
use warnings;
use CGI;
use DBI;
use JSON;
use Audio::TagLib;

my $cgi = CGI->new();
my $mp3Root = "/music";
my $imgRoot = "/music/.covers";
my $dbConn;

# set of (pl, gn, ar, al, sn)
# search of
#   gn ->
#   ar -> tags, note, genre
#   al -> artist, genre, year, tags, note
#   sn -> title, track, tags, note, ar, al, year, gn
#   pl -> title, tags, note, ar, al, sn, year, gn
#   ps -> plTitle

reqLibrary();

sub reqLibrary {
  # my @params = $cgi->param();
  if(!defined $cgi->param('qtype')) { textPlain(""); }
  if($cgi->param('qtype') eq 'total' ) { reqTotal (); }
  if($cgi->param('qtype') eq 'genre' ) { reqGenre (); }
  if($cgi->param('qtype') eq 'artist') { reqArtist(); }
  if($cgi->param('qtype') eq 'album' ) { reqAlbum (); }
  if($cgi->param('qtype') eq 'song'  ) { reqSong  (); }
}

sub reqTotal {
  my $q;
  my %obj;
  $dbConn = dbGet();
  $q = dbQuery("SELECT COUNT(*) FROM song"  ); $q->execute(); $obj{'songs'  } = $q->fetchrow_arrayref()->[0]; $q->finish();
  $q = dbQuery("SELECT COUNT(*) FROM album" ); $q->execute(); $obj{'albums' } = $q->fetchrow_arrayref()->[0]; # $q->finish();
  $q = dbQuery("SELECT COUNT(*) FROM artist"); $q->execute(); $obj{'artists'} = $q->fetchrow_arrayref()->[0]; $q->finish();
  textPlain(encode_json(\%obj));
}

sub reqGenre {
  $dbConn = dbGet();
  my $query = dbQuery("SELECT DISTINCT genre FROM artist"); $query->execute();
  my @arr;
  for(my $i=0; $i<$query->rows; $i++) {
    push(@arr, $query->fetchrow_arrayref()->[0]);
  }
  $query->finish();
  textPlain(encode_json(\@arr));
}

sub reqArtist {
  $dbConn = dbGet();
  my $query = "SELECT indexArtist, name, genre, tags, note FROM artist"; my $where=""; my @params;
  if(defined $cgi->param('tags'))  { $where.=" AND " if($where ne ""); $where.="(tags LIKE ?)"; push(@params, '%'. $cgi->param('tags') .'%'); }
  if(defined $cgi->param('note'))  { $where.=" AND " if($where ne ""); $where.="(note LIKE ?)"; push(@params, '%'. $cgi->param('note') .'%'); }
  if(defined $cgi->param('genre')) { $where.=" AND " if($where ne ""); $where.="(genre = ?)"  ; push(@params, $cgi->param('genre'));          }
  $query .= " WHERE " . $where if($where ne "");
  $query = dbQuery($query); $query->execute(@params);
  my @arr;
  for(my $i=0; $i<$query->rows; $i++) {
    my %obj; my $ref = $query->fetchrow_arrayref();
    $obj{'indexArtist'} = $ref->[0];
    $obj{'name'       } = $ref->[1];
    $obj{'genre'      } = $ref->[2];
    $obj{'tags'       } = $ref->[3];
    $obj{'note'       } = $ref->[4];
    push(@arr, \%obj);
  }
  $query->finish();
  textPlain(encode_json(\@arr));
}

sub reqAlbum {
  $dbConn = dbGet(); my @params;
  my $query = "SELECT artist.name, album.title, album.year, album.cover, album.tags, album.note";
  my $from  = " FROM album, artist";
  my $where = " WHERE (artist.indexArtist = album.indexArtist)";
  if(defined $cgi->param('artist')) { $where.=" AND " if($where ne ""); $where.="(artist.name = ?)"  ; push(@params, $cgi->param('artist')); }
  if(defined $cgi->param('genre' )) { $where.=" AND " if($where ne ""); $where.="(artist.genre = ?)" ; push(@params, $cgi->param('genre' )); }
  if(defined $cgi->param('year'  )) { $where.=" AND " if($where ne ""); $where.="(album.year = ?)"   ; push(@params, $cgi->param('year'  )); }
  if(defined $cgi->param('tags'  )) { $where.=" AND " if($where ne ""); $where.="(album.tags LIKE ?)"; push(@params, '%'. $cgi->param('tags') .'%'); }
  if(defined $cgi->param('note'  )) { $where.=" AND " if($where ne ""); $where.="(album.note LIKE ?)"; push(@params, '%'. $cgi->param('note') .'%'); }
  $query .= $from if($from ne "");
  $query .= $where if($where ne "");
  $query = dbQuery($query); $query->execute(@params);
  my @arr;
  for(my $i=0; $i<$query->rows; $i++) {
    my %obj; my $ref = $query->fetchrow_arrayref();
    $obj{'artist'    } = $ref->[0];
    $obj{'title'     } = $ref->[1];
    $obj{'year'      } = $ref->[2];
    $obj{'cover'     } = $ref->[3];
    $obj{'tags'      } = $ref->[4];
    $obj{'note'      } = $ref->[5];
    push(@arr, \%obj);
  }
  $query->finish();
  textPlain(encode_json(\@arr));
}

sub reqSong {
  $dbConn = dbGet(); my @params;
  my $query = "SELECT song.indexSong, artist.name, album.title, song.title, song.track, song.tags, song.note, album.year, artist.genre, song.file";
  my $from  = " FROM song, album, artist";
  my $where = " WHERE (song.indexAlbum = album.indexAlbum) AND (song.indexArtist = artist.indexArtist)";
  if(defined $cgi->param('index' )) { $where.=" AND " if($where ne ""); $where.="(song.indexSong = ?)"; push(@params, $cgi->param('index' )); }
  if(defined $cgi->param('artist')) { $where.=" AND " if($where ne ""); $where.="(artist.name = ?)"   ; push(@params, $cgi->param('artist')); }
  if(defined $cgi->param('album' )) { $where.=" AND " if($where ne ""); $where.="(album.title = ?)"   ; push(@params, $cgi->param('album' )); }
  if(defined $cgi->param('genre' )) { $where.=" AND " if($where ne ""); $where.="(artist.genre = ?)"  ; push(@params, $cgi->param('genre' )); }
  if(defined $cgi->param('year'  )) { $where.=" AND " if($where ne ""); $where.="(album.year = ?)"    ; push(@params, $cgi->param('year'  )); }
  if(defined $cgi->param('tags'  )) { $where.=" AND " if($where ne ""); $where.="(song.tags LIKE ?)"  ; push(@params, '%'. $cgi->param('tags' ) .'%'); }
  if(defined $cgi->param('note'  )) { $where.=" AND " if($where ne ""); $where.="(song.note LIKE ?)"  ; push(@params, '%'. $cgi->param('note' ) .'%'); }
  if(defined $cgi->param('title' )) { $where.=" AND " if($where ne ""); $where.="(song.title LIKE ?)" ; push(@params, '%'. $cgi->param('title') .'%'); }
  $query .= $from if($from ne "");
  $query .= $where if($where ne "");
  $query = dbQuery($query); $query->execute(@params);
  my @arr;
  for(my $i=0; $i<$query->rows; $i++) {
    my %obj; my $ref = $query->fetchrow_arrayref();
    $obj{'indexSong'} = $ref->[0];
    $obj{'artist'   } = $ref->[1];
    $obj{'album'    } = $ref->[2];
    $obj{'title'    } = $ref->[3];
    $obj{'track'    } = $ref->[4];
    $obj{'tags'     } = $ref->[5];
    $obj{'note'     } = $ref->[6];
    $obj{'year'     } = $ref->[7];
    $obj{'genre'    } = $ref->[8];
    $obj{'file'     } = $ref->[9];
    push(@arr, \%obj);
  }
  $query->finish();
  textPlain(encode_json(\@arr));
}

sub textPlain {
  my $content = shift;
  print "Content type: text/plain\r\n\r\n";
  print $content;
  exit 0;
}

sub dbQuery {
  my $id3=3;
  return $dbConn->prepare(shift);
}

sub dbGet {
  my $src  = "DBI:mysql:database=musicapp;host=localhost";
  my $conn = DBI->connect($src, "perldb", "----------"); # insert password
  
  return $conn;
}

File: app/py/library.py
import json
import pcgi as cgi
import pymysql as sql

# ==============================================================================
def request():
    if not cgi.has_param('qtype'):
        cgi.response('')
        return
    qtype = cgi.get_param('qtype')
    if   qtype == 'total' : request_total()
    elif qtype == 'genre' : request_genre()
    elif qtype == 'artist': request_artist()
    elif qtype == 'album' : request_album()
    elif qtype == 'song'  : request_song()
    else: cgi.response_404()

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def request_total():
    songs   = db_query('SELECT COUNT(*) FROM song')[0][0]
    albums  = db_query('SELECT COUNT(*) FROM album')[0][0]
    artists = db_query('SELECT COUNT(*) FROM artist')[0][0]
    cgi.response_object(songs=songs, albums=albums, artists=artists)

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def request_genre():
    genres = db_query('SELECT DISTINCT genre FROM artist')
    genres = [x[0] for x in genres]
    cgi.response(json.dumps(genres))

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def request_artist():
    query = 'SELECT indexArtist, name, genre, tags, note FROM artist'
    where = []
    params = []

    if cgi.has_param('genre'):
        where.append('(genre = ?)')
        params.append(cgi.get_param('genre'))
    if cgi.has_param('tags'):
        where.append('(tags LIKE ?)')
        params.append(''.join(['%', cgi.get_param('tags'), '%']))
    if cgi.has_param('note'):
        where.append('(note LIKE ?)')
        params.append(''.join(['%', cgi.get_param('note'), '%']))

    if len(where):
        where = ' AND '.join(where)
        query = ' WHERE '.join([query, where])
    if not len(params):
        params = None

    arr = []
    for item in db_query(query, params=params):
        arr.append({'indexArtist' : item[0],
                    'name' : item[1],
                    'genre': item[2],
                    'tags' : item[3],
                    'note' : item[4]})
    cgi.response(json.dumps(arr))

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def request_album():
    query = 'SELECT album.indexAlbum, artist.name, album.title, album.year, album.cover, album.tags, album.note, artist.genre FROM album, artist'
    where = ['(artist.indexArtist = album.indexArtist)']
    params = []

    if cgi.has_param('artist'):
        where.append('(artist.name = ?)')
        params.append(cgi.get_param('artist'))
    if cgi.has_param('genre'):
        where.append('(artist.genre = ?)')
        params.append(cgi.get_param('genre'))
    if cgi.has_param('year'):
        where.append('(album.year = ?)')
        params.append(cgi.get_param('year'))
    if cgi.has_param('tags'):
        where.append('(album.tags LIKE ?)')
        params.append(''.join(['%', cgi.get_param('tags'), '%']))
    if cgi.has_param('album.note'):
        where.append('(note LIKE ?)')
        params.append(''.join(['%', cgi.get_param('note'), '%']))

    if len(where):
        where = ' AND '.join(where)
        query = ' WHERE '.join([query, where])
    if not len(params):
        params = None

    arr = []
    for item in db_query(query, params=params):
        arr.append({'indexAlbum' : item[0],
                    'artist' : item[1],
                    'title'  : item[2],
                    'year'   : item[3],
                    'cover'  : item[4],
                    'tags'   : item[5],
                    'note'   : item[6],
                    'genre'  : item[7]})
    cgi.response(json.dumps(arr))

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def request_song():
    query = 'SELECT song.indexSong, artist.name, album.title, song.title, song.track, song.tags, song.note, album.year, artist.genre, song.file FROM song, album, artist'
    where = ['(song.indexAlbum = album.indexAlbum)', '(song.indexArtist = artist.indexArtist)']
    params = []

    if cgi.has_param('indexAlbum'):
        where.append('(song.indexAlbum = ?)')
        params.append(cgi.get_param('indexAlbum'))
    if cgi.has_param('index'):
        where.append('(song.indexSong = ?)')
        params.append(cgi.get_param('index'))
    if cgi.has_param('artist'):
        where.append('(artist.name = ?)')
        params.append(cgi.get_param('artist'))
    if cgi.has_param('album'):
        where.append('(album.title = ?)')
        params.append(cgi.get_param('album'))
    if cgi.has_param('genre'):
        where.append('(artist.genre = ?)')
        params.append(cgi.get_param('genre'))
    if cgi.has_param('year'):
        where.append('(album.year = ?)')
        params.append(cgi.get_param('year'))
    if cgi.has_param('tags'):
        where.append('(songs.tags LIKE ?)')
        params.append(''.join(['%', cgi.get_param('tags'), '%']))
    if cgi.has_param('note'):
        where.append('(songs.note LIKE ?)')
        params.append(''.join(['%', cgi.get_param('note'), '%']))
    if cgi.has_param('title'):
        where.append('(song.title LIKE ?)')
        params.append(''.join(['%', cgi.get_param('title'), '%']))

    if len(where):
        where = ' AND '.join(where)
        query = ' WHERE '.join([query, where])
    if not len(params):
        params = None

    arr = []
    for item in db_query(query, params=params):
        arr.append({'indexSong' : item[0],
                    'artist' : item[1],
                    'album'  : item[2],
                    'title'  : item[3],
                    'track'  : item[4],
                    'tags'   : item[5],
                    'note'   : item[6],
                    'year'   : item[7],
                    'genre'  : item[8],
                    'file'   : item[9]})
    cgi.response(json.dumps(arr))

# ==============================================================================
_DB = None
def db_get():
    global _DB
    if _DB is None:
        _DB = sql.connect(host='localhost', user='USER', passwd='PASS', db='musicapp')
        db_query('SET NAMES utf8')
    return _DB

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def db_close():
    global _DB
    if _DB is not None:
        _DB.close()

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

def db_query(query, params=None):
    db = db_get()
    if params is not None:
        params = [db.escape(x) for x in params]
        query = query.replace('?','{}').format(*params)
    cur = db.cursor()
    cur.execute(query)
    result = cur.fetchall() or []
    cur.close()
    return result

# ==============================================================================
cgi.init()
request()
db_close()

# ==============================================================================
# ------------------------------------------------------------------------------
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
